README.md
## About

This exporter intented to expose metrics based on Elasticsearch queries.
This exporter never run in production, so use it on your own risk :-)

## Requirements

Any modern Linux distro (Arch, Ubuntu, CentOS, Debian).

Following `python` libraries are required: `pyyaml`, `requests`, `prometheus_client`.

## Configuration File

Configuration file is in a yaml format with single configuration scope (`metrics`).

Each item in `metrics` scope define a metric and must have following attributes:

- `metric_desc`: description of a metric (what it does)
- `metric_value`: reference to value in Elasticsearch response (json data)
- `metric_labels`: inner scope with `metric_name: reference` for each metric:
    - `metric_name`: name of label exposed by exporter
    - `reference`: reference to value in Elasticsearch response (json data)
- `es_url`: url to Elasticsearch cluster (include index)
- `es_query`: query in json format; must
    - _either_ be inclosed in single quotes (e.g. `'{ "query": {...} }'`)
    - _or_ written in [YAML block notation](http://yaml.org/spec/1.2/spec.html#|%20literal%20style//) with proper indentation, e.g.,

        ```
        es_query: |
          {
            "query": {
              ...
            }
          }
        ```

Refer to `exporter_es.yml` as an example.

## Installation and running

```
virtualenv .pyenv
. .pyenv/bin/activate
pip install pyyaml requests prometheus_client
python exporter_es.py --conf exporter_es.yml --log-level INFO --log-file /tmp/exporter_es.log
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.3] - 2018-10-26

### Added

- travis-ci configuration
- Explicit support for Python 3.5

### Fixed

- Added missing __init__.py

## [0.0.2] - 2018-07-04

### Changed

- Change default port from 9100 to 9349 to be a good OSS citizen and use a free port from the to-be-considered
  [Prometheus port range](https://github.com/prometheus/prometheus/wiki/Default-port-allocations)

## [0.0.1] - 2018-07-04

### Added

- Basic configurable exporter, gathering data from Elasticsearch and reformatting it in
  [Prometheus' exposition format](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md)

[Unreleased]: https://github.com/croesnick/promesque/compare/v0.0.2...HEAD
[0.0.2]: https://github.com/croesnick/promesque/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/croesnick/promesque/compare/v0.0.1


promesque/cli.py
import logging
import sys
import time

import click
from prometheus_client import start_http_server
from ruamel import yaml

from promesque.lib.exporter import ESGaugeMetric
from promesque.lib.exporter_logger import ExporterLogger, LOG_LEVEL_MAP

logger = ExporterLogger('promesque')


@click.command()
@click.argument('config', type=click.Path(exists=True))
@click.option('--port', default=9349, help='Port to bind prometheus exporter')
@click.option('--interval', default=60, help='Interval to probe nodes and update metrics (in seconds)')
@click.option('--log-file', type=click.Path(writable=True), help='Path to log file to write')
@click.option('--log-level', type=click.Choice(LOG_LEVEL_MAP.keys()), default='error')
def cli(config, port, interval, log_file, log_level):
    logger.setLevel(ExporterLogger.level(log_level))
    if log_file:
        file_handler = logging.FileHandler(log_file)
        file_handler.setFormatter(logger.formatter)
        logger.addHandler(file_handler)

    with open(config, 'r') as f:
        try:
            config = yaml.safe_load(f)
        except Exception as e:
            logger.error('Failed to parse configuration file {0}. Errors: {1}'.format(config, e))
            sys.exit(1)

    metrics = []

    conf_metrics = config['metrics']
    for name, properties in conf_metrics.items():
        url = properties['url']
        description = properties['description']
        data_path = properties.get('data_path', '$')
        value_path = properties['value_path']
        query = properties['query']

        labels = {}
        if 'labels' in properties:
            labels = properties['labels']

        metric = ESGaugeMetric(name, description, labels, data_path, value_path, url, query, logger=logger)
        metrics.append(metric)

    start_http_server(port)

    while True:
        for metric in metrics:
            try:
                metric.update(print_metric=True)
            except Exception as e:
                logger.error(e)
        time.sleep(interval)


if __name__ == '__main__':
    cli()


promesque/lib/exporter.py
import json

import requests
from jsonpath_ng import jsonpath, parse
from prometheus_client import Gauge


class ESGaugeMetric(object):
    def __init__(self, name, description, labels, data_path, value_path, url, query, value_converter=None, logger=None):
        '''
        name            -- name of metric (e.g., node_network_status)
        description     -- description of metric
        labels          -- indexes (tuple of strings) in metric_data taken as labels
        data_path       -- jsonpath to buckets of data to build metrics from
        value_path      -- jsonpath to value of the metric, located inside data referred to by `data_path`
        value_converter -- sometime value may came in mixed format like - 5s, 3GB.
                           we need to convert this value to numeric.
                           pass a function reference to this converter, can be lambda as well.
        url             -- elasticsearch url to index or GET query
        query           -- elasticsearch query data for POST request
        logger          -- instance of logging.Logger class
        '''
        self.gauge = Gauge(name, description, list(labels))
        self.name = name
        self.labels = labels
        self.data_path = data_path
        self.value_path = value_path
        self.value_converter = value_converter
        self.url = url
        self.query = query
        self.logger = logger

    @classmethod
    def es_query(cls, url, data):
        '''
        query Elasticsearch cluster and return raw requests.Response object
        url     -- url to elastic search e.g. - http://localhost:9200/bank/_search
        data    -- query in json format - more info reffer to Elasticsearch documentation
        return  -- raw requests.Response object
        '''
        headers = {'Content-Type': 'application/json'}
        resp = requests.post(url, headers=headers, data=data)
        return resp

    def metrics(self, metric_data):
        '''
        populate labels and value with data
        metric_data -- dict object
        return      -- labels - list of dicts with label=value, value
        '''
        value_jsonpath = parse(self.value_path)

        data_jsonpath = parse(self.data_path)
        data = [item.value for item in data_jsonpath.find(metric_data)]

        for item in data:
            value_matches = value_jsonpath.find(item)
            if len(value_matches) != 1:
                self.logger.error('Expected exactly one metric value at value path={0}, got {1}'.format(
                    self.value_path,
                    len(value_matches)))
                raise ValueError

            value = value_matches[0].value
            if self.value_converter:
                value = self.value_converter(value)

            labels = {}
            for name, path in self.labels.items():
                label_jsonpath = parse(path)
                label_matches = label_jsonpath.find(item)
                if len(label_matches) != 1:
                    self.logger.error('Expected exactly one label value at path={0}, got {1}'.format(
                        path,
                        len(label_matches)))
                    raise ValueError

                labels[name] = label_matches[0].value

            yield labels, value

    def print_metric(self, metric_labels, metric_value):
        '''
        build and print metric
        metric_labels -- labels to print
        metric_value  -- value to print
        '''
        if metric_labels:
            label_value = []
            for label, value in metric_labels.items():
                label_value.append('{l}="{v}"'.format(l=label, v=value))
            # show labels in a log
            text = '{n}{{{lv}}} {v}'.format(n=self.name, lv=', '.join(label_value), v=metric_value)
        else:
            # there are no labels to show
            text = '{n} {v}'.format(n=self.name, v=metric_value)
        if self.logger:
            self.logger.info(text)
        else:
            print('[INFO]: {t}'.format(t=text))
        
    def update(self, print_metric=False):
        '''
        query ES and update metric with newer value
        print_metric    -- print metric to stdout (good for dev stage)
        '''
        resp = self.es_query(self.url, data=self.query)
        data = json.loads(resp.text)
        for labels, value in self.metrics(data):
            if print_metric:
                self.print_metric(labels, value)
            if self.labels:
                self.gauge.labels(**labels).set(value)
            else:
                self.gauge.set(value)


promesque/lib/exporter_logger.py
import logging
import sys


LOG_LEVEL_MAP = {
    'debug': logging.DEBUG,
    'info': logging.INFO,
    'warning': logging.WARNING,
    'error': logging.ERROR,
    'critical': logging.CRITICAL,
}


class ExporterError(Exception):
    pass


class ExporterLogger(logging.Logger):
    def __init__(self, name, path=None, level='error', fmt='%(asctime)s [%(levelname)-5.5s]:  %(message)s'):
        self._path = path
        self._level = self.level(level)

        super(ExporterLogger, self).__init__(name, self._level)
        self._formatter = logging.Formatter(fmt)

        stream_handler = logging.StreamHandler(sys.stdout)
        stream_handler.setFormatter(self._formatter)
        self.addHandler(stream_handler)

    @classmethod
    def level(cls, level):
        return LOG_LEVEL_MAP[level.lower()]

    @property
    def formatter(self):
        return self._formatter


exporter_es.py
import os
import sys
import json
import time
import logging
import optparse
import requests
import yaml
from prometheus_client import start_http_server, Gauge

class ExporterError(Exception):
    pass

class ExporterLogger(logging.Logger):
    def __init__(self, name, path=None, level='ERROR', fmt='%(asctime)s [%(levelname)-5.5s]:  %(message)s'):
        try:
            # resolve textual level to numerical 
            level = getattr(logging, level)
        except Exception:
            raise ExporterError('wrong logging level {l}'.format(l=level))
        super(ExporterLogger, self).__init__(name, level)
        formatter = logging.Formatter(fmt)
        stream_handler = logging.StreamHandler(sys.stdout)
        stream_handler.setLevel(level)
        stream_handler.setFormatter(formatter)
        self.addHandler(stream_handler)
        if path:
            file_handler = logging.FileHandler(path)
            file_handler.setLevel(level)
            file_handler.setFormatter(formatter)
            self.addHandler(file_handler)

class ESGaugeMetric(object):
    def __init__(self, name, desc, labels, value, value_converter, url, query, logger=None):
        '''
        name            -- metric name (e.g. node_network_status)
        desc            -- metric description
        labels          -- indexes (tuple of strings) in metric_data taken as labels
        value           -- index in metric_data (dict) taken as value for metric
        value_converter -- sometime value may came in mixed format like - 5s, 3GB.
                           we need to convert this value to numeric.
                           pass a function reference to this converter, can be lambda as well.
        url             -- elasticsearch url to index or GET query
        query           -- elasticsearch query data for POST request
        logger          -- instance of logging.Logger class
        '''
        self.gauge = Gauge(name, desc, list(labels))
        self.name = name
        self.labels = labels
        self.value = value
        self.value_converter = value_converter
        self.url = url
        self.query = query
        self.logger = logger

    def path_converter(self, path):
        '''
        convert path from indexA.indexB to ['indexA']['indexB']
        path    -- path in dot notaion
        return  -- path in bracket notaion
        '''
        elems = []
        for elem in path.split('.'):
            bracket = "['{0}']".format(elem)
            elems.append(bracket)
        return ''.join(elems)

    def es_query(self, url, data):
        '''
        query Elasticsearch cluster and return raw requests.Response object
        url     -- url to elastic search e.g. - http://localhost:9200/bank/_search
        data    -- query in json format - more info reffer to Elasticsearch documentation
        return  -- raw requests.Response object
        '''
        headers = {'Content-Type': 'application/json'}
        resp = requests.post(url, headers=headers, data=data)
        return resp

    def populate(self, metric_data):
        '''
        populate labels and value with data
        metric_data -- dict object
        return      -- metric_labels - dict with label=value, metric_value - converted value
        '''
        try:
            converter = getattr(self, self.value_converter)
        except Exception:
            converter = self.value_converter
        value_path = self.path_converter(self.value)
        value_var = 'metric_data{0}'.format(value_path)
        metric_value = converter(eval(value_var))
        metric_labels = {}
        for label_name, label_path in self.labels.items():
            label_path = self.path_converter(label_path)
            label_var = 'metric_data{0}'.format(label_path)
            metric_labels[label_name] = eval(label_var)
        return metric_labels, metric_value

    def print_metric(self, metric_labels, metric_value):
        '''
        build and print metric
        metric_labels -- labels to print
        metric_value  -- value to print
        '''
        if metric_labels:
            label_value = []
            for label, value in metric_labels.items():
                label_value.append('{l}={v}'.format(l=label, v=value))
            # show labels in a log
            text = '{n}{{{lv}}} {v}'.format(n=self.name, lv=', '.join(label_value), v=metric_value)
        else:
            # there are no labels to show
            text = '{n} {v}'.format(n=self.name, v=metric_value)
        if self.logger:
            self.logger.info(text)
        else:
            print('[INFO]: {t}'.format(t=text))
        
    def update(self, print_metric=False):
        '''
        query ES and update metric with newer value
        print_metric    -- print metric to stdout (good for dev stage)
        '''
        resp = self.es_query(self.url, data=self.query)
        metric_data = json.loads(resp.text)
        metric_labels, metric_value = self.populate(metric_data)
        if print_metric:
            self.print_metric(metric_labels, metric_value)
        if self.labels:
            self.gauge.labels(**metric_labels).set(metric_value)
        else:
            self.gauge.set(metric_value)

if __name__ == '__main__':
    usage = '%prog --conf [--port --interval, --log-file, --log-level]'
    parser = optparse.OptionParser(usage=usage)
    parser.add_option('-p', '--port', type='int', help='port to bind prometheus exporter', default=9100)
    parser.add_option('-i', '--interval', type='int', help='interval to probe nodes and update metrics', default=60)
    parser.add_option('-c', '--conf', type='str', help='path to a conf file')
    parser.add_option('-l', '--log-file', type='str', help='path to a log file', default=None)
    parser.add_option('-L', '--log-level', type='str', help='levels available: DEBUG,INFO,WARNING,ERROR,CRITICAL', default='ERROR')
    opts, args = parser.parse_args()
    logger = ExporterLogger('exporter_es', path=opts.log_file, level=opts.log_level)
    if not opts.conf:
        raise ExporterError('configuration file is a required argument')
    if not os.path.isfile(opts.conf):
        raise ExporterError('file {c} do not exists'.format(c=opts.conf))
    with open(opts.conf, 'r') as f:
        try:
            conf = yaml.load(f)
        except Exception as ex:
            print('[ERROR]: failed to parse configuration file {0} ...'.format(opts.conf))
            sys.exit(1)
    # store all metrics in a list
    metrics = []
    # get metric params from config
    conf_metrics = conf['metrics']
    for metric_name, metric_params in conf_metrics.items():
        metric_desc = metric_params['metric_desc']
        metric_value = metric_params['metric_value']
        es_url = metric_params['es_url']
        es_query = metric_params['es_query']
        # metric_labels is not required config param, hence need to create it's default
        metric_labels = {}
        if 'metric_labels' in metric_params:
            metric_labels = metric_params['metric_labels']
        # create and store metric
        metric = ESGaugeMetric(metric_name, metric_desc, metric_labels, metric_value, int, es_url, es_query, logger=logger)
        metrics.append(metric)
    start_http_server(opts.port)
    while True:
        for metric in metrics:
            try:
                metric.update(print_metric=True)
            except Exception as ex:
                logger.error(ex)
        time.sleep(opts.interval)
